domain/storage/service/storagepool.go

```diff
@@ -0,0 +1,246 @@
+	sp := domainstorage.StoragePoolDetails{
+		Name:     name,
+		Provider: string(providerType),
+		Attrs:    attrsToSave,
+	}
+	err = s.st.ReplaceStoragePool(ctx, sp)
+	return errors.Annotatef(err, "replacing storage pool %q", name)
+}
+
+// ListStoragePools returns the storage pools matching the specified filter.
+func (s *StoragePoolService) ListStoragePools(ctx context.Context, filter domainstorage.StoragePoolFilter) ([]*storage.Config, error) {
+	if err := s.validatePoolListFilter(filter); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	pools, err := domainstorage.BuiltInStoragePools()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	sp, err := s.st.ListStoragePools(ctx, filter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	pools = append(pools, sp...)
+
+	results := make([]*storage.Config, len(pools))
+	for i, p := range pools {
+		results[i], err = s.toStorageConfig(p)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return results, nil
+}
+
+func (s *StoragePoolService) validatePoolListFilter(filter domainstorage.StoragePoolFilter) error {
+	if err := s.validateProviderCriteria(filter.Providers); err != nil {
+		return errors.Trace(err)
+	}
+	if err := s.validateNameCriteria(filter.Names); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (s *StoragePoolService) validateNameCriteria(names []string) error {
+	for _, n := range names {
+		if !storage.IsValidPoolName(n) {
+			return fmt.Errorf("pool name %q not valid%w", n, errors.Hide(storageerrors.InvalidPoolNameError))
+		}
+	}
+	return nil
+}
+
+func (s *StoragePoolService) validateProviderCriteria(providers []string) error {
+	if s.registry == nil {
+		return errors.New("cannot filter storage providers without a registry")
+	}
+	for _, p := range providers {
+		_, err := s.registry.StorageProvider(storage.ProviderType(p))
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// GetStoragePoolByName returns the storage pool with the specified name, returning an error
+// satisfying [storageerrors.PoolNotFoundError] if it doesn't exist.
+func (s *StoragePoolService) GetStoragePoolByName(ctx context.Context, name string) (*storage.Config, error) {
+	if !storage.IsValidPoolName(name) {
+		return nil, fmt.Errorf("pool name %q not valid%w", name, errors.Hide(storageerrors.InvalidPoolNameError))
+	}
+
+	builtIn, err := domainstorage.BuiltInStoragePools()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, p := range builtIn {
+		if p.Name == name {
+			return s.toStorageConfig(p)
+		}
+	}
+	sp, err := s.st.GetStoragePoolByName(ctx, name)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return s.toStorageConfig(sp)
+}
+
+func (s *StoragePoolService) toStorageConfig(sp domainstorage.StoragePoolDetails) (*storage.Config, error) {
+	if s.registry == nil {
+		return nil, errors.New("cannot load storage pools without a registry")
+	}
```

Comment:
When could this happen? I think we should remove this. If there is a chance that you might not have a registry then either pass in a NoopRegistry or re-architect this service to be two services. One called `StoragePoolService` and `StoragePoolRegistryService`, then any calls to the registry have to be only accessed by the latter.

Reply:
This can happen because the service constructors off the model service factory conventionally do not return an error.

So we have:

```
// Storage returns the model's storage service.
func (s *ModelFactory) Storage(registry storage.ProviderRegistry) *storageservice.Service {
	return storageservice.NewService(
		storagestate.NewState(changestream.NewTxnRunnerFactory(s.modelDB)),
		s.logger.Child("storage"),
		registry,
	)
}
```
```
func NewService(st State, logger Logger, registry storage.ProviderRegistry) *Service {
	return &Service{
		StoragePoolService: &StoragePoolService{
			st:       st,
			logger:   logger,
			registry: registry,
		},
	}
}
```

We do pass in a registry but there's no convention to return an error if the registry is nil. If we want to allow model factory methods to return an error, then we can check for nil there.